src/lib/rules/strict-transport-security/strict-transport-security.ts

```diff
@@ -0,0 +1,128 @@
+// ------------------------------------------------------------------------------
+
+const rule: IRuleBuilder = {
+    create(context: RuleContext): IRule {
+        /** The minimum period (in seconds) allowed for `max-age`. */
+        let minMaxAgeValue: number;
+
+        /** match for each directive in the header */
+        type DirectiveMatch = { // eslint-disable-line no-unused-vars
+            'max-age': Array<string>,
+            'includesubdomains': Array<string>,
+            'preload': Array<string>
+        };
+
+        const loadRuleConfigs = () => {
+            minMaxAgeValue = (context.ruleOptions && context.ruleOptions.minMaxAgeValue) || 10886400; // 18 weeks
+        };
+
+        // STS header Syntax:
+        // Strict-Transport-Security: max-age=<expire-time>
+        // Strict-Transport-Security: max-age=<expire-time>; includeSubDomains
+        // Strict-Transport-Security: max-age=<expire-time>; preload
+        // This function accomplishes the following:
+        // "max-age=31536000; includesubdomains; preload" => {"max-age":31536000,"includesubdomains":true,"preload":true}
+        const parse = (headerValue: string) => {
+            const parsedHeader = {};
+            const directives = headerValue.toLowerCase().split(';');
+            const nameValuePairRegex = /^ *([!#$%&'*+.^_`|~0-9A-Za-z-]+) *= *("(?:[~0-9])*"|[!#$%&'*+.^_`|~0-9]+) *$/;
+            // Regex for name-value pairs. E.g.: max-age=31536000
+            // Modified usage of https://github.com/jshttp/content-type/blob/64bde0d996ccb4334341662c0c7d25f7b370c4d9/index.js#L23
+            const tokenRegex = /^ *[!#$%&'*+.^_`|~0-9A-Za-z-]+$/;
+            // Regex for single tokens. E.g.:  includesubdomains
+
+            directives.forEach((directive) => {
+                const match = tokenRegex.exec(directive) || nameValuePairRegex.exec(directive);
+
+                if (!match) {
+                    throw new Error(`'strict-transport-security' header has the wrong format: ${directive}`);
+                }
+
+                const [matchString, key, value] = match;
+                const name = key || matchString.trim();
+
+                if (parsedHeader[name]) {
+                    throw new Error(`'strict-transport-security' header contains more than one '${name}'`);
+                }
+
+                parsedHeader[name] = value || true;
+            });
+
+            return parsedHeader;
+        };
+
+        const isInvalidValue = (maxAge: string, limit: number): boolean => {
+            return maxAge && parseInt(maxAge) < limit;
+        };
+
+        const validate = async (fetchEnd: IFetchEnd) => {
+            const { element, resource, response }: { element: IAsyncHTMLElement, resource: string, response: IResponse } = fetchEnd;
+            const headerValue: string = normalizeString(response.headers && response.headers['strict-transport-security']);
+            let parsedHeader;
+
+            if (!isHTTPS(resource)) {
```

Comment:
> So that means the header shouldn't be required if the url before redirect is http, right?

Yeap, and that is because over HTTP an attacker may intercept the connections and remove the header. 

Talking with @qzhou1607 it seems my comment was not clear. What I suggested was that even if the user provides a `HTTP` URL that, for some reason is not redirected to `HTTPS`,  we should still check if the site supports `HTTPS`, and if it does check for the header for `HTTPS`.

----

Basically: 

  * if it's `HTTP` - the presence of the header should be an error
  * if it's `HTTPS` - the lack of the header should be an error

----

> Add preloadList with a default to false to check for those values:

@qzhou1607 We should make it very clear in to docs that users should not do it if they are not 100% sure they can support `HTTPS` for their entire site and all its subdomains in the long run (see [my initial comment](https://github.com/sonarwhal/sonar/issues/23#issue-213013019)).

> Should we download the list and see if the website has been already submitted?

@molant Do you mean as in: do the tests as if `preloadList` was set to `true` if we encounter a site from that list?